coding-articles/yehia-article-thegraph/Article.md

```diff
@@ -0,0 +1,660 @@
+
+```
+// PairCreated is an event inside the smart contract
+export function handlePairCreated(event: PairCreated): void {
+    // What will happen after the event is triggered
+}
+```
+
+### YAML file
+
+`subgraph.yaml` map handlers to the event. For example, to make the above handler work, we will link it with the event inside subgraph.yml this way
+
+```
+eventHandlers:
+        - event: PairCreated(indexed address,indexed address,address,uint256)
+          handler: handlePairCreated
+  file: ./<mapper file location>/map.ts
+```
+
+Now you understand the three primary files to create a subgraph; you can start by installing the graph-cli
+
+```
+yarn global add @graphprotocol/graph-cli
+```
+
+Create a directory and initialize the project
+
+```
+# Inside the project directory
+graph init
+```
+
+This command will ask you some questions
+
+1. Protocol? Ethereum
+2. Product for which to initialize? Subgraph-studio hosted service may be better if you want to control who queries your graph and the cost
+3. Subgraph slug? I will name it swappairs
+4. Ethereum Network? mainnet
+5. Contract Address? 0x5c69bee701ef814a2b6a3edd4b1652cb9cc5aa6f
+6. Contract Name? Pairs
+   Then the project will start installing, and you should see something like this 👇
+
+<figure>
+  <img src="./images/graph_init.png" alt="Initialize graph output">
+</figure>
+
+After installing, you should now access the project from your code editor.
+
+<figure>
+  <img src="./images/project_root.png" alt="Initialize graph output">
+</figure>
+
+Start with the entities inside schema.graphql and put the data you need to
+First schema is the `Token` it will be simple we need only the contract address and the token name
+
+```
+type Token @entity {
+  id: ID! # contract address
+  name: String!
+}
+```
+
+Second schema is the `CreatedPairs` its the two pair of tokens details, the transaction hash and the timestamp.
+
+```
+type CreatedPair @entity {
+  id: ID! # transaction hash
+  token0: Token!
+  token1: Token!
+  timestamp: BigInt!
+}
+```
+
+Schema file is ready you can check it on [src/schema.graphql](code/subgraph/schema.graphql) last thing is to generate the schema entities by running
+
+```
+yarn codegen
+```
+
+command I will explain it in details later.
+
+Lets do the logic now that
+Go to [src/swap-pairs.ts](code/subgraph/schema.graphql) delete everything inside the file
+
+Import the schema file
+
+```
+// Import the created swap entity
+import { PairCreated } from "../generated/SwapPairs/SwapPairs";
+```
+
+Create the first handler
+
+```
+// PairCreated is an event inside the smart contract
+export function handlePairCreated(event: PairCreated): void {
+    // What will happen after the event is triggered
+}
+```
+
+Inside the `handlePairCreated` function lets initialize the entity for `CreatedPair`
+
+```
+  // Inside handlePairCreated
+  let createdPair = new CreatedPair(event.transaction.hash.toHexString());
```

Comment:
I think this will work only if we first import CreatedPair from schema, right?

Reply:
yess exactly; I forget most of the imports because Vscode detect them automatically 😅  I updated it, sorry for that